File: frontend/src/services/api/locations.ts
import { apiRequest } from "./client";

export type LocationOpenStatus = "open" | "closed_temporarily" | "closed_permanently";

/**
 * Where a profile came from. `fixture` is the sample dataset the API serves while Google
 * has not approved Business Profile API access for this project, and it is labelled as
 * such everywhere it is shown.
 */
export type LocationSource = "google" | "fixture" | "manual";

export interface LocationSummary {
  id: string;
  title: string;
  /** Prebuilt single-line address from the API; null when Google holds no address. */
  address: string | null;
  primary_category_display: string | null;
  open_status: LocationOpenStatus | null;
  has_voice_of_merchant: boolean;
  has_pending_edits: boolean;
  has_google_updated: boolean;
  is_duplicate: boolean;
  source: LocationSource;
  store_code: string | null;
  last_synced_at: string | null;
}

export interface LocationCategory {
  category_name: string;
  display_name: string | null;
  is_primary: boolean;
}

export interface LocationHoursPeriod {
  hours_type: string;
  open_day: string;
  open_hour: number;
  open_minute: number;
  close_day: string;
  close_hour: number;
  close_minute: number;
}

export interface LocationAttribute {
  attribute_id: string;
  value_type: string;
  values: unknown[];
}

export interface AttributeCatalogItem {
  external_attribute_id: string;
  attribute_name: string;
  attribute_group: string;
  applies_to_category: string;
  value_type: string;
}

export interface AttributeCatalog {
  items: AttributeCatalogItem[];
  total: number;
}

export interface LocationDetail extends LocationSummary {
  google_location_name: string;
  google_resource_name: string | null;
  source_location_id: string | null;
  place_id: string | null;
  address_lines: string[] | null;
  locality: string | null;
  administrative_area: string | null;
  postal_code: string | null;
  region_code: string | null;
  latitude: number | null;
  longitude: number | null;
  phone_primary: string | null;
  website_uri: string | null;
  description: string | null;
  opening_date: string | null;
  maps_uri: string | null;
  new_review_uri: string | null;
  created_at: string | null;
  updated_at: string | null;
  categories: LocationCategory[];
  hours_periods: LocationHoursPeriod[];
  attributes: LocationAttribute[];
}

/**
 * An edit submitted for review. Every field is optional: only the ones present are
 * considered, and an explicit `null` clears the value on the Google profile.
 */
export interface LocationEditRequest {
  title?: string;
  phone_primary?: string | null;
  website_uri?: string | null;
  description?: string | null;
  open_status?: LocationOpenStatus;
  hours_periods?: LocationHoursPeriod[];
}

/** One before/after pair, already rendered as display strings by the API. */
/**
 * One before/after pair. Scalar fields arrive as strings, but structured ones — hours
 * periods, attributes — arrive as arrays of objects so the client can format them
 * legibly rather than showing raw JSON.
 */
export interface FieldChange {
  field: string;
  label: string;
  current: unknown;
  proposed: unknown;
}

/**
 * The result of dry-running an edit against Google. `changes` is the only diff the
 * product shows; the client never computes its own.
 */
export interface EditPreview {
  changes: FieldChange[];
  update_mask: string[];
  /** Keyed by the request field name, for example `title` or `website_uri`. */
  field_errors: Record<string, string>;
  valid: boolean;
}

export type ProfileActionStatus =
  | "pending"
  | "approved"
  | "executing"
  | "succeeded"
  | "failed";

/** One entry of the audit trail for a profile. */
export interface ProfileAction {
  id: string;
  action_type: string;
  status: ProfileActionStatus;
  payload: Record<string, unknown>;
  error: string | null;
  created_at: string;
  /** The person who submitted the action; null when the API does not identify one. */
  user: string | null;
}

/** The audit endpoint returns a bare array, newest first — not a paged envelope. */
export type ProfileActionList = ProfileAction[];

/** The API caps a page at 200 locations, so that is what the list surfaces request. */
export const LOCATIONS_MAX_PAGE_SIZE = 200;

export interface LocationListParams {
  /** Restrict the result to the locations that belong to one project. */
  projectId?: string;
  /** Server-side search across the location fields. */
  q?: string;
  /** Defaults to 50 on the API; hard capped at 200. */
  limit?: number;
  offset?: number;
}

function locationPath(id: string) {
  return `/locations/${encodeURIComponent(id)}`;
}

export const locationsApi = {
  list: ({ projectId, q, limit, offset }: LocationListParams = {}) => {
    const query = new URLSearchParams();
    if (projectId) query.set("project_id", projectId);
    if (q) query.set("q", q);
    if (limit !== undefined) query.set("limit", String(limit));
    if (offset !== undefined) query.set("offset", String(offset));
    const search = query.toString();
    return apiRequest<LocationSummary[]>(`/locations${search ? `?${search}` : ""}`);
  },
  get: (id: string) => apiRequest<LocationDetail>(locationPath(id)),
  attributeCatalog: () => apiRequest<AttributeCatalog>("/locations/attribute-catalog"),
  /**
   * Dry-runs the edit against Google. Nothing on the live listing changes, so this is
   * safe to call as often as the person editing wants to look at the diff.
   */
  previewEdit: (id: string, input: LocationEditRequest) =>
    apiRequest<EditPreview>(`${locationPath(id)}/edits/preview`, {
      method: "POST",
      body: JSON.stringify(input),
    }),
  /** Applies the edit to the live Google listing and records it in the audit trail. */
  applyEdit: (id: string, input: LocationEditRequest) =>
    apiRequest<ProfileAction>(`${locationPath(id)}/edits`, {
      method: "POST",
      body: JSON.stringify(input),
    }),
  listActions: (id: string) => apiRequest<ProfileActionList>(`${locationPath(id)}/actions`),
};

File: backend/app/schemas/locations.py
from datetime import date, datetime
from typing import Annotated, Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models import ActionStatus, LocationSource, OpenStatus


class LocationCategoryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    category_name: str
    display_name: str | None = None
    is_primary: bool = False


class LocationHoursPeriodResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    hours_type: str = "REGULAR"
    open_day: str
    open_hour: int
    open_minute: int = 0
    close_day: str
    close_hour: int
    close_minute: int = 0


class LocationAttributeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    attribute_id: str
    value_type: str
    values: Any = None


class AttributeCatalogItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    external_attribute_id: str
    attribute_name: str
    attribute_group: str
    applies_to_category: str
    value_type: str


class AttributeCatalogResponse(BaseModel):
    items: list[AttributeCatalogItemResponse] = []
    total: int = 0


class LocationSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    google_location_name: str
    title: str
    store_code: str | None = None
    primary_category_display: str | None = None
    address: str | None = None
    locality: str | None = None
    administrative_area: str | None = None
    open_status: OpenStatus | None = None
    has_voice_of_merchant: bool = False
    has_pending_edits: bool = False
    has_google_updated: bool = False
    is_duplicate: bool = False
    # Where this profile came from. `fixture` is the sample dataset served while Google
    # has not approved API access, and the UI labels it as such on every row.
    source: LocationSource = LocationSource.google
    last_synced_at: datetime | None = None


class LocationDetail(LocationSummary):
    model_config = ConfigDict(from_attributes=True)

    google_resource_name: str | None = None
    source_location_id: str | None = None
    place_id: str | None = None
    primary_category_name: str | None = None
    address_lines: list[str] | None = None
    postal_code: str | None = None
    region_code: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    phone_primary: str | None = None
    website_uri: str | None = None
    description: str | None = None
    opening_date: date | None = None
    maps_uri: str | None = None
    new_review_uri: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    categories: list[LocationCategoryResponse] = []
    hours_periods: list[LocationHoursPeriodResponse] = []
    attributes: list[LocationAttributeResponse] = []


# ---------------------------------------------------------------------------
# Editing a profile
# ---------------------------------------------------------------------------


class HoursPeriodInput(BaseModel):
    """One opening period. Sending `hours_periods` replaces the whole regular schedule,
    which is what Google's `regularHours` field does."""

    open_day: str
    open_hour: Annotated[int, Field(ge=0, le=23)]
    open_minute: Annotated[int, Field(ge=0, le=59)] = 0
    close_day: str
    close_hour: Annotated[int, Field(ge=0, le=24)]
    close_minute: Annotated[int, Field(ge=0, le=59)] = 0


class AttributeInput(BaseModel):
    attribute_id: str
    value_type: str = "BOOL"
    values: list[Any] = []


class CategoryInput(BaseModel):
    category_name: str
    display_name: str | None = None


class LocationEditRequest(BaseModel):
    """Every field is optional, and *omitted* is meaningfully different from *null*.

    Only the fields actually present in the request body are considered; the diff against
    the stored location decides what ends up in the update mask. A client cannot ask for a
    field it did not send to be touched, which is what keeps an unrelated field from being
    unset by a patch.
    """

    title: str | None = None
    phone_primary: str | None = None
    website_uri: str | None = None
    description: str | None = None
    open_status: OpenStatus | None = None
    hours_periods: list[HoursPeriodInput] | None = None
    attributes: list[AttributeInput] | None = None
    primary_category: CategoryInput | None = None


class FieldChange(BaseModel):
    field: str
    label: str
    current: Any = None
    proposed: Any = None


class EditPreviewResponse(BaseModel):
    """The dry run. `valid` is true only when there is something to send and Google
    accepted the request with `validateOnly=true`."""

    changes: list[FieldChange] = []
    update_mask: list[str] = []
    field_errors: dict[str, str] = {}
    valid: bool = False


class ActionUser(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    email: str
    full_name: str | None = None


class ProfileActionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    action_type: str
    status: ActionStatus
    payload: Any = None
    error: str | None = None
    created_at: datetime
    user: ActionUser | None = None
